CONTRIBUTING.md

```diff
@@ -1,73 +1,76 @@
-
-### Naming convention for gtest macros:
-
-TEST(ModuleNameClassNameTest, TestName), e.g., TEST(OperatorsGetTableTest, RowCount)
-same for fixtures Test_F()
-
-If you want to test a single module, class or test you have to execute the test binary and use the `gtest_filter` option:
-
-- Testing the storage module: `./build/hyriseTest --gtest_filter="Storage*"`
-- Testing the table class: `./build/hyriseTest --gtest_filter="StorageTableTest*"`
-- Testing the RowCount test: `./build/hyriseTest --gtest_filter="StorageTableTest.RowCount"`
-
-## Performance Warnings
-- Sometimes, we have convenience functions, such as `ReferenceSegment::operator[]`, or workarounds, such as performing multiple stable sorts instead of a single one. Because these might negatively affect the performance, the user should be warned if a query causes one of these slow paths to be chosen. For this, we have the PerformanceWarning() macro defined in `performance_warning.hpp`.
-
-## Documentation
-- Most documentation should happen in the code or in the beginning of the header file
-- More complex documentation, such as an explanation of an algorithm that profits from images, can be put in the Wiki. Please make sure to link the Wiki page in the code - otherwise, no one will find it.
-
-# Review
-
-**Things to look for:**
-
-	- Guidelines (see above)
-	- Is the copy constructor deleted where it makes sense?
-	- Is the destructor virtual for base classes?
-	- Are unnecessary copies of heavy elements made? (prefer vector& over vector, but not int& over int)
-	- Did the author update documentation and dependencies (Wiki, README.md, DEPENDENCIES.md, Dockerfile, install_dependencies.sh, Jenkinsfile)
+# The Four Commandments
+1. Thy code (or the beginning of the header file) shalt be the primary method of documentation. Comments should be used to explain higher level concepts, but primarily, the code should be explaining itself. Part of this is to choose concise, descriptive names.
+1. Thou shalt program defensively. Use Assert wherever it makes sense, use DebugAssert in performance-critical parts (e.g., within hot loops). Also, we do not handle exceptions in Hyrise. If an invalid state is reached, we crash immediately. This makes debugging easier. An exception (hihi) to this is user-facing code where we handle, e.g., typos in SQL queries.
+1. Thou shalt reflect on existing code. Just as your code is not perfect, neither is the code that people wrote before you. Try to improve it as part of your PRs and do not hesitate to ask if anything is unclear. Chances are that it can be improved.
+1. Thou shalt properly test thy code. This includes unit *and* integration tests. Try to isolate parts that can be individually tested.
+
+# C++
+* Use automatic memory management (RAII, smart pointers). `new` and `malloc` are evil words.
```

Comment:
Since we had a ton of (unnecessary) shared_ptr's in the sort merge join, I am wondering if there good guidelines by Herb and co. when to use them and when not. I haven't seen this overuse lately though.

Reply:
Is this something that can be boiled down into a proper guideline though? I don't think you can teach the concepts of ownership in this document. I'm happy to add

> Be mindful of ownership. Not everything needs to be a smart pointer. Consider passing around references to the object or references to a shared_ptr instead of copying the shared_ptr. Remember that this might not be safe when passing shared_ptrs into, i.e., JobTasks.

Anything beyond that is, in my opinion, the job of the seminar.

Reply:
That's fine with me. I link would probably help though (I don't like any of the links above tbh ... Herb has a post, but that one does not reason about when to use pointers at all ... https://herbsutter.com/2013/05/29/gotw-89-solution-smart-pointers/)